adaptiveRouter.go

```diff
@@ -0,0 +1,100 @@
+package solr
+
+import (
+	"net/http"
+	"sort"
+	"sync"
+	"time"
+)
+
+type adaptive []*searchHistory
+
+func (s adaptive) Len() int {
+	return len(s)
+}
+
+func (s adaptive) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (s adaptive) Less(i, j int) bool {
+	return s[i].getErrors() < s[j].getErrors() && s[i].getAvgLatency() < s[j].getAvgLatency()
+}
+
+type adaptiveRouter struct {
+	history map[string]*searchHistory
+	recency int
+	lock    *sync.RWMutex
+}
+
+func (q *adaptiveRouter) SortUris(urisIn []string) []string {
+	q.lock.RLock()
+	defer q.lock.RUnlock()
+	searchHistory := make(adaptive, len(urisIn))
+	for i, uri := range urisIn {
+		if v, ok := q.history[uri]; ok {
+			searchHistory[i] = v
+		} else {
+			searchHistory[i] = newLatencyHistory(uri, q.recency)
+		}
+	}
+
+	sort.Sort(searchHistory)
+	sorted := make([]string, len(searchHistory))
+	for i, uri := range searchHistory {
+		sorted[i] = uri.uri
+	}
+	return sorted
+}
+
+func (q *adaptiveRouter) AddSearchResult(t time.Duration, uri string, resp *http.Response, err error) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	if _, ok := q.history[uri]; !ok {
+		q.history[uri] = newLatencyHistory(uri, q.recency)
+	}
+	success := (err != nil && resp.StatusCode == http.StatusOK)
+	q.history[uri].addSearchResult(t, !success)
+}
+
+type searchHistory struct {
+	timings []time.Duration
+	errors  []bool
+	uri     string
+	offset  int
+}
+
+func newLatencyHistory(uri string, recency int) *searchHistory {
+	return &searchHistory{uri: uri, timings: make([]time.Duration, recency), errors: make([]bool, recency)}
+}
+
+func (u *searchHistory) addSearchResult(timing time.Duration, error bool) {
+	u.timings[u.offset] = timing
+	u.errors[u.offset] = error
+	u.offset++
+	if u.offset == len(u.timings) {
```

Comment:
Better to instantiate 0 length with N capacity, and then use cap(u.timings) for the check.